### ISSUE TITLE
Gaussian wavenet how many steps for good convergence

### ISSUE THREAD

**Issue by G-Wang** [Rayhane-mamah/Tacotron-2#229]
Hi, any one have experience with how many steps it takes for good quality gaussian wavenet with batch size of 8 (the default hparams values).

**Comment by Rayhane-mamah** [comment]
Hey @G-Wang sorry for the late reply!

From my personal experience on different speakers/datasets, it takes between 500k and 750k steps to converge with batch_size=8 and utterance length of 0.5s (depending on your sample rate). 
I am also not noticing any particular benefit from using WeightNorm for the time being on the 20-layers model, so one can omit them for now.

**Comment by G-Wang** [comment]
Thanks for the reply, and just wanted to say that you have one of the top quality TTS repo on github, and I have learned alot from it, so thank you for your open source effort!

I'm working on a different vocoder (WaveRNN alternative: https://github.com/fatchord/WaveRNN), and I was having trouble with single gaussian output.

I then updated it from Gaussian to a Beta distribution (which has bounded support between 0 and 1), and it seems to train much better (I did tweak with the learning rate schedule as well so that could be the reason too)

Lastly I don't have a tensorflow set up atm so I can't test your code, but do you have rough ball-park numbers for the Wavenet generation rate (samples/second)?

I saw Nvidia's real-time wavenet generation models (with their custom CUDA code), however it seems to only support 256 mu-law encoding.

**Comment by Rayhane-mamah** [comment]
thanks for your support @G-Wang! 

Ah! fatchord got himself a nice piece of work as well on his side :)

As for your generation speed question: the stats I am providing are using a (12GB) Titan XP  for graph operations so it may differ for other users:
- 8-bit WaveNet (mulaw encoding): ~310 samples/sec
- 16-bit WaveNet (Gaussian/MoL): ~240 samples/sec

Now those are the average generation rates for a **single utterance synthesis** (made faster than original WaveNet 172sample/sec by following [fast wavenet](https://arxiv.org/pdf/1611.09482.pdf) algorithms during synthesis). In this repo, we do however support batch synthesis on WaveNet which typically provides a sub-linear acceleration until the GPU computation capacity hits 100% (parallel computation is bigger than the capacity of the GPU so they are not entirely run at the same time). After that, small speed boosts are due to effective memory transfer in and out of GPU memory. I typically found that with a Titan XP one can use batch_size=20 (depending on the lengths of the utterances) without problems making wavenet synthesis close to 6~8x slower than real time when processing batches (single utterance synthesis is ~60x slower than real time).

Finally, nv-wavenet indeed is a great intelligent management of the GPU resources to accelerate the WaveNet model. I personally prefer their persistent kernels over the two other approaches, but it usually requires not very large models. While that work focuses on DeepVoice 1 WaveNet variant, it can be adapted for other WaveNet variants (like our 20-layer Gaussian WaveNet inspired by Baidu ClariNet) I believe, provided one is proficient with Cuda. :) 

Also considering that WaveRNN is a relatively small model (1 masked GRU + few FC projections) in contrast to WaveNet (multiple dilated residual blocks and skip convolutions), the persistent kernel implementation can work perfectly with WaveRNN I believe. It still requires quite some effort to make the implementation however I believe.

If you ever do make a "nv-WaveRNN" implementation, know that I will be your bigger fan :)

**Comment by G-Wang** [comment]
By batching do you mean breaking the features (Mel) for a single utterance into batches, forward pass through model then stich them back into a single wav at the end?

**Comment by tdplaza** [comment]
My results at 164k steps


[164000_steps.zip](https://github.com/Rayhane-mamah/Tacotron-2/files/2444997/164000_steps.zip)
[originalVoice.zip](https://github.com/Rayhane-mamah/Tacotron-2/files/2444998/originalVoice.zip)
